pwndbg/emu/emulator.py

```diff
@@ -528,6 +528,25 @@ def update_pc(self, pc=None) -> None:
             pc = pwndbg.gdblib.regs.pc
         self.uc.reg_write(self.get_reg_enum(self.regs.pc), pc)
 
+    def read_thumb_bit(self) -> int | None:
+        """
+        Return 0 or 1, representing the status of the Thumb bit in the current Arm architecture
+
+        This reads from the emulator itself, meaning this can be read to determine a state
+        transitions between non-Thumb and Thumb mode
+
+        Return None if the Thumb bit is not relevent to the current architecture
+
+        Mimics the `read_thumb_bit` function defined in gdblib/arch.py
+        """
+        if self.arch == "arm":
+            if (cpsr := self.cpsr) is not None:
+                return (cpsr >> 5) & 1
+        elif self.arch == "armcm":
+            if (xpsr := self.xpsr) is not None:
+                return (xpsr >> 24) & 1
```

Comment:
I am okay with this code but I think we should just use a correct version/register based on known architecture.

I mean, let's analyze this.

1. The `self.arch` is set in the `Emulator` object based on current arch figured out by Pwndbg: https://github.com/pwndbg/pwndbg/blob/b74fcdde97a4cd4c6c072235040eb21ee7a42a5b/pwndbg/emu/emulator.py#L184-L188
2. The `pwndbg.gdblib.arch.current` is set in the `Arch` class here: https://github.com/pwndbg/pwndbg/blob/b74fcdde97a4cd4c6c072235040eb21ee7a42a5b/pwndbg/lib/arch.py#L16-L21
3. This is updated by the `gdblib.arch.update` function: https://github.com/pwndbg/pwndbg/blob/b74fcdde97a4cd4c6c072235040eb21ee7a42a5b/pwndbg/gdblib/arch.py#L109-L111
4. Which is triggered as a hook here: https://github.com/pwndbg/pwndbg/blob/b74fcdde97a4cd4c6c072235040eb21ee7a42a5b/pwndbg/gdblib/hooks.py#L22-L26

I wonder:
- Do we really need to call it on each stop too? I bet we do it since the `stop` event may be called while none other event is called? Shouldn't we cache it if the arch did not change or sth?
- If we figure out the arch, maybe its time to do some "architecture classes" and have a `def get_thumb_bit` method defined differently for each arch? I mean, we do call such function anyway, so why not define this function differently for each arch when we know this arch already?

Reply:
I completely agree. The codebase has a couple (5+) occurences of long `if/else` chains that check for the current architecture, and set a local variable/take some small actions based on the arch.

I've put some notes on the Pwndbg Discord regarding a new architecture class structure, where we can have functions such as `get_thumb_bit` which would reduce the amount of these `if/else` checks, and at the same time bundling in some other changes to cleanup the use of `ABI's`.

Once the architecture changes are in, we can come back and fix these occurences of `if arch == "arm"`, etc, etc.

A note on the need for updating the arch on each step:
- I can think of at least one case where it is necessary, which is the early stages of a bootloader in x86 when it changes from 16-bit mode to 32-bit instruction set.